cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/runtime/spi/provisioner/TetheringProvisioner.java

```diff
@@ -0,0 +1,120 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.tethering.runtime.spi.provisioner;
+
+import com.google.inject.Guice;
+import com.google.inject.Injector;
+import com.google.inject.PrivateModule;
+import com.google.inject.Scopes;
+import io.cdap.cdap.api.metrics.MetricsCollectionService;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.guice.ConfigModule;
+import io.cdap.cdap.common.guice.InMemoryDiscoveryModule;
+import io.cdap.cdap.internal.tethering.runtime.spi.runtimejob.TetheringRuntimeJobManager;
+import io.cdap.cdap.messaging.MessagingService;
+import io.cdap.cdap.messaging.guice.MessagingServerRuntimeModule;
+import io.cdap.cdap.metrics.collect.LocalMetricsCollectionService;
+import io.cdap.cdap.runtime.spi.provisioner.Capabilities;
+import io.cdap.cdap.runtime.spi.provisioner.Cluster;
+import io.cdap.cdap.runtime.spi.provisioner.ClusterStatus;
+import io.cdap.cdap.runtime.spi.provisioner.PollingStrategies;
+import io.cdap.cdap.runtime.spi.provisioner.PollingStrategy;
+import io.cdap.cdap.runtime.spi.provisioner.Provisioner;
+import io.cdap.cdap.runtime.spi.provisioner.ProvisionerContext;
+import io.cdap.cdap.runtime.spi.provisioner.ProvisionerSpecification;
+import io.cdap.cdap.runtime.spi.runtimejob.RuntimeJobManager;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A provisioner that does not create or tear down clusters, but just uses an existing tethered CDAP instance.
+ */
+public class TetheringProvisioner implements Provisioner {
+
+  public static final String TETHERING_NAME = "tethering";
+  private static final ProvisionerSpecification SPEC = new ProvisionerSpecification(
+    TETHERING_NAME, "Tethering Provisioner",
+    "Runs programs on an existing tethered CDAP instance. Does not provision any resources.");
+
+  @Override
+  public ProvisionerSpecification getSpec() {
+    return SPEC;
+  }
+
+  @Override
+  public void validateProperties(Map<String, String> properties) {
+    // Creates the TetheringConf for validation
+    TetheringConf.fromProperties(properties);
+  }
+
+  @Override
+  public Cluster createCluster(ProvisionerContext context) {
+    return new Cluster(context.getProgramRunInfo().getRun(), ClusterStatus.RUNNING,
+                       Collections.emptyList(), Collections.emptyMap());
+  }
+
+  @Override
+  public ClusterStatus getClusterStatus(ProvisionerContext context, Cluster cluster) {
+    ClusterStatus status = cluster.getStatus();
+    return status == ClusterStatus.DELETING ? ClusterStatus.NOT_EXISTS : status;
```

Comment:
when would the cluster status be DELETING?

Reply:
[Provisioner interface](https://github.com/cdapio/cdap/blob/348b579c316c4ff74dab6ea6d663be046ff85cad/cdap-runtime-spi/src/main/java/io/cdap/cdap/runtime/spi/provisioner/Provisioner.java#L118-L144) has a default `deleteClusterWithStatus()` method (that modifies the cluster status to DELETING) that is called by [ClusterDeleteSubtask](https://github.com/cdapio/cdap/blob/348b579c316c4ff74dab6ea6d663be046ff85cad/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/task/ClusterDeleteSubtask.java)